src/coreclr/vm/interpexec.cpp

```diff
@@ -1493,6 +1548,121 @@ do {                                                                           \
                     STELEM(double, double);
                     break;
                 }
+                case INTOP_STELEM_REF:
+                {
+                    BASEARRAYREF arrayRef = LOCAL_VAR(ip[1], BASEARRAYREF);
+                    if (arrayRef == NULL)
+                        COMPlusThrow(kNullReferenceException);
+
+                    ArrayBase* arr = (ArrayBase*)OBJECTREFToObject(arrayRef);
+                    uint32_t len = arr->GetNumComponents();
+                    uint32_t idx = (uint32_t)LOCAL_VAR(ip[2], int32_t);
+                    if (idx >= len)
+                        COMPlusThrow(kIndexOutOfRangeException);
+
+                    OBJECTREF elemRef = LOCAL_VAR(ip[3], OBJECTREF);
+
+                    if (elemRef != NULL)
+                    {
+                        MethodTable* arrayElemMT = (MethodTable*)pMethod->pDataItems[ip[4]];
+                        if (arrayElemMT != NULL && !ObjIsInstanceOf(OBJECTREFToObject(elemRef), arrayElemMT))
```

Comment:
```suggestion
                        TypeHandle arrayElemType = arr->GetArrayElementTypeHandle();
                        if (!ObjIsInstanceOf(OBJECTREFToObject(elemRef), arrayElemType))
```
This check should be done against the actual array type (different from LDELEMA_REF).